gramex/data.py

```diff
@@ -1391,28 +1392,66 @@ def _type_conversion(param_list, operations=None):
 }
 
 
-def _filter_mongodb_col(col, op, vals, meta_cols):
+def _filter_mongodb_col(col, op, vals, results, object_keys=[]):
+    from dateutil.parser import parse
+
+    if not results:
+        return
+
     if op in ['', '!']:
-        return {col: {_mongodb_op_map[op]: vals}}
+        convert = type(results[0][col])
+
+        if convert == datetime:
+            convert = parse
+
+        if convert == bool:
+            for index, val in enumerate(vals):
+                vals[index] = 0 if val in ['0', 'false', 'False'] else 1
+
+        return {col: {_mongodb_op_map[op]: [convert(val) for val in vals]}}
+    elif op in ['>', '>~', '<', '<~']:
+        convert = type(results[0][col])
+
+        if convert == datetime:
+            convert = parse
+
+        vals = [convert(val) for val in vals if val]
+
+        if op == '>':
+            return {col: {_mongodb_op_map[op]: max(vals)}}
+        elif op == '<':
+            return {col: {_mongodb_op_map[op]: min(vals)}}
+        elif op == '>~':
+            return {col: {_mongodb_op_map[op]: max(vals)}}
+        elif op == '<~':
+            return {col: {_mongodb_op_map[op]: min(vals)}}
     elif op == '!~':
         return {col: {"$not": {"$regex": '|'.join(vals), "$options": 'i'}}}
     elif op == '~':
         return {col: {"$regex": '|'.join(vals), "$options": 'i'}}
     elif col and op in _mongodb_op_map.keys():
         # TODO: Improve the numpy to Python type
-        convert = int if (meta_cols[col].dtype == pd.np.int64) else meta_cols[col].dtype.type
+        convert = type(results[0][col])
         return {col: {_mongodb_op_map[op]: convert(val)} for val in vals}
 
 
-def _mongodb_query(args, meta_cols):
+def _mongodb_query(args, meta_cols, results=None, **kwargs):
```

Comment:
Don't pass results here. Instead, take table. Set `results = table.find().limit(1)` -- we're not using more rows. Handle scenario where there are no rows, i.e. no data